src/Http/Http.Extensions/src/JsonOptions.cs

```diff
@@ -21,11 +23,21 @@ public class JsonOptions
         // Because these options are for producing content that is written directly to the request
         // (and not embedded in an HTML page for example), we can use UnsafeRelaxedJsonEscaping.
         Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
+
+        // The JsonSerializerOptions.GetTypeInfo method is called directly and needs a defined resolver
+        // setting the default resolver (reflection-based) but the user can overwrite it directly or calling
+        // .AddContext<TContext>()
+        TypeInfoResolver = CreateDefaultTypeResolver()
     };
 
     // Use a copy so the defaults are not modified.
     /// <summary>
     /// Gets the <see cref="JsonSerializerOptions"/>.
     /// </summary>
     public JsonSerializerOptions SerializerOptions { get; } = new JsonSerializerOptions(DefaultSerializerOptions);
+
+    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026",
+        Justification = "The default type resolver will be set as DefaultJsonTypeInfoResolver (reflection-based) that has the same runtime behavior when TypeResolver is null.")]
+    private static IJsonTypeInfoResolver? CreateDefaultTypeResolver()
```

Comment:
Even with a features switch, having this suppression in JsonTypeInfo class itself feels bad. Doesn't this mean that someone could now use the initial JsonOptions.SerializerOptions value to manually get a reflection-based JsonTypeInfo from DefaultJsonTypeInfoResolver that's not really trim-safe and then pass it to one the "safe" JsonTypeInfo serialize overloads yourself and get zero warnings because of this suppression?

Arguably the DefaultJsonTypeInfoResolver.GetTypeInfo method should have warnings since that's doing the reflection, but that's suppressed because the constructor has the warnings. If we're going to suppress this, we should ensure that only we can use this resolver from code that is properly attributed with `RequiresUnreferencedCode` and `RequiresDynamicCode`.

Reply:
We should only be setting this default in the non AOT case, right?

Reply:
> We should only be setting this default in the non AOT case, right?

Yeah, I could add this in this pr but I decided to have another one focused on AOT work.

Reply:
> Even with a features switch, having this suppression in JsonOptions class itself feels bad. This could be used from anywhere. Doesn't this mean that someone could now use the initial JsonOptions.SerializerOptions value to manually get a reflection-based JsonTypeInfo from DefaultJsonTypeInfoResolver that's not really trim-safe and then pass it to one the "safe" JsonTypeInfo serialize overloads yourself and get zero warnings because of this suppression?

After I started working on AOT part of this behavior I am sure the `Suppress` is wrong and even adding `RequiresUnReferenceCode` is not what we want. My suggestion is add #pragma here to not break our build and this will warning every time for users., until my PR to add a feature switch that will trim the `new DefaultJsonTypeInfoResolver() ` part a not throw warnings anymore where the feature is `on` that will be the default when PublishAOT or Publish Trimmed.

Reply:
> will warning every tike for users 

I mean every time the analyzer is enabled.

Reply:
It is not a good idea to suppress linker warnings unless there is reasonable justification that every potential callsite is already annotated as RUC (one common example that jumps to mind is suppressing instance methods in a class whose constructors have been annotated as RUC). RUC annotations should bubble all the way up to public APIs (roughly speaking, this is similar to how we avoid doing sync over async and always surface async calls to public APIs).

For this particular case, I would recommend marking the `DefaultSerializerOptions` property in line 19 as RUC/RDC.

cc @eerhardt

Reply:
I have a discussion about using RUC, while work on the feature switch, and it is not correct since the trial is Trimmer will no be able to optimize at the point of avoid the warning. That is why my suggestion is using just #pragma to Suppress in out build.

Reply:
Just to make it clear, here is my suggestion:

``` c#
    public static readonly JsonSerializerOptions DefaultSerializerOptions = new JsonSerializerOptions(JsonSerializerDefaults.Web)
    {
#pragma warning disable IL2026 // Members annotated with 'RequiresUnreferencedCodeAttribute' require  
        TypeInfoResolver = JsonTrimmability.IsTrimmable ? null : CreateDefaultResolver()
#pragma warning restore IL2026 // Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code
    };

    [RequiresUnreferencedCode("Calls System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.DefaultJsonTypeInfoResolver()")]
    private static IJsonTypeInfoResolver CreateDefaultResolver() => new DefaultJsonTypeInfoResolver();
```

This part ` JsonTrimmability.IsTrimmable ? null :` will come in my next PR. Make sense?

Reply:
While a flag _could_ prevent runtime errors related to trimmed code from occurring, generally speaking linker-safety is a static attribute of an application, rather than a dynamic one. In other words, trimmed/AOT applications calling into linker-unsafe code need to provide warnings at compile time, rather than depending on runtime configuration to call into the right branch. Suppressing warnings for this particular public property means that certain applications won't produce any compile-time warnings even though the produced artifact contains linker/AOT-unsafe code.

Reply:
The `IsTrimmable` is a feature switch and will no warning in AOT. Also, the suggestion is not suppressing the warning and instead warning everything when the feature switch is not `on`